src/system.rs

```diff
@@ -142,105 +156,485 @@ impl StageExecutor {
+// so the underlying Accessor type functions from the queries don't need to allocate.
+// Otherwise, this leads to excessive alloaction for every call to reads/writes
+/// The core builder of `System` types, which are systems within Legion. Systems are implemented
+/// as singular closures for a given system - providing queries which should be cached for that
+/// system, as well as resource access and other metadata.
+/// ```rust
+/// # use legion::prelude::*;
+/// # use legion::resource::ResourceAccessType;
+/// # #[derive(Copy, Clone, Debug, PartialEq)]
+/// # struct Position;
+/// # #[derive(Copy, Clone, Debug, PartialEq)]
+/// # struct Velocity;
+/// # #[derive(Copy, Clone, Debug, PartialEq)]
+/// # struct Model;
+/// #[derive(Copy, Clone, Debug, PartialEq)]
+/// struct Static;
+///
+///  let mut system_one = SystemBuilder::<()>::new("TestSystem")
+///            .with_resource::<TestResource>(ResourceAccessType::Read)
+///            .with_query(<(Read<Position>, Tagged<Model>)>::query()
+///                         .filter(!tag::<Static>() | changed::<Position>()))
+///            .build(move |_resource, queries| {
+///                println!("Hello world");
+///               let mut count = 0;
+///                {
+///                    for (entity, pos) in queries.iter_entities() {
+///                        assert_eq!(expected.get(&entity).unwrap().0, *pos);
+///                        count += 1;
+///                    }
+///                }
+///
+///                assert_eq!(components.len(), count);
+///            });
+/// ```
+struct SystemBuilder<Q = (), R = ()> {
+    name: String,
+
+    queries: Q,
+    resources: R,
+
+    resource_access: Access<TypeId>,
+    component_access: Access<ComponentTypeId>,
+}
+
+impl<Q, R> SystemBuilder<Q, R>
+where
+    Q: 'static + Send + ConsFlatten,
+    R: 'static + Send + ConsFlatten,
+{
+    #[allow(clippy::new_ret_no_self)]
+    fn new(name: &str) -> SystemBuilder {
+        SystemBuilder {
+            name: name.to_string(),
+            queries: (),
+            resources: (),
+            resource_access: Access::default(),
+            component_access: Access::default(),
+        }
+    }
+
+    pub fn with_query<V, F>(
+        mut self,
+        query: Query<V, F>,
+    ) -> SystemBuilder<<Q as ConsAppend<Query<V, F>>>::Output, R>
+    where
+        V: for<'a> View<'a>,
+        F: 'static + EntityFilter,
+        Q: ConsAppend<Query<V, F>>,
+    {
+        self.component_access.reads.extend(V::read_types().iter());
+        self.component_access.writes.extend(V::write_types().iter());
+
+        SystemBuilder {
+            name: self.name,
+            queries: ConsAppend::append(self.queries, query),
+            resources: self.resources,
+            resource_access: self.resource_access,
+            component_access: self.component_access,
+        }
+    }
+
+    fn with_resource<T>(
```

Comment:
Similar to an earlier comment, the resource accesses should ultimately compile into `&T` and `&mut T` references to be passed to the system function in the resource tuple. This is surely going to require that `with_resource` knows statically whether the resource is going to be read or written?